src/writer/stripe.rs

```diff
@@ -0,0 +1,175 @@
+        // Direct encoding to represent root struct
+        let mut column_encodings = vec![ColumnEncoding::Direct];
+        let child_column_encodings = self
+            .columns
+            .iter()
+            .map(|c| c.column_encoding())
+            .collect::<Vec<_>>();
+        column_encodings.extend(child_column_encodings);
+        let column_encodings = column_encodings.iter().map(From::from).collect();
+
+        // Root type won't have any streams
+        let mut written_streams = vec![];
+        for (index, c) in self.columns.iter_mut().enumerate() {
+            // Offset by 1 to account for root of 0
+            let column = index + 1;
+            let streams = c.finish();
+            // Flush the streams to the writer
+            for s in streams {
+                let (s_type, bytes) = s.into_parts();
+                let length = bytes.len();
+                self.writer.write_all(&bytes).context(IoSnafu)?;
+                written_streams.push(WrittenStream {
+                    s_type,
+                    column,
+                    length,
+                });
+            }
+        }
+        let streams = written_streams.iter().map(From::from).collect();
+        let stripe_footer = proto::StripeFooter {
+            streams,
+            columns: column_encodings,
+            writer_timezone: None,
+            encryption: vec![],
+        };
+
+        let data_length = self.writer.stream_position().context(IoSnafu)? - start_offset;
+        let footer_bytes = stripe_footer.encode_to_vec();
+        let footer_length = footer_bytes.len() as u64;
+        let row_count = self.row_count;
+        self.writer.write_all(&footer_bytes).context(IoSnafu)?;
+
+        // Reset state for next stripe
+        self.row_count = 0;
+
+        Ok(StripeInformation {
+            start_offset,
+            index_length: 0,
+            data_length,
+            footer_length,
+            row_count,
+        })
+    }
+
+    /// Close this writer and return the inner writer.
+    pub fn close(self) -> W {
+        self.writer
+    }
+}
+
+fn create_encoder(field: &FieldRef) -> Box<dyn ColumnStripeEncoder> {
+    match field.data_type() {
+        ArrowDataType::Float32 => Box::new(FloatStripeEncoder::new(ColumnEncoding::Direct)),
+        ArrowDataType::Float64 => Box::new(DoubleStripeEncoder::new(ColumnEncoding::Direct)),
+        ArrowDataType::Int8 => Box::new(ByteStripeEncoder::new(ColumnEncoding::Direct)),
+        ArrowDataType::Int16 => Box::new(Int16StripeEncoder::new(ColumnEncoding::DirectV2)),
+        ArrowDataType::Int32 => Box::new(Int32StripeEncoder::new(ColumnEncoding::DirectV2)),
+        ArrowDataType::Int64 => Box::new(Int64StripeEncoder::new(ColumnEncoding::DirectV2)),
+        // TODO: support more datatypes
+        _ => unimplemented!("unsupported datatype"),
+    }
+}
+
+#[derive(Copy, Clone, Debug, Eq, PartialEq)]
+struct WrittenStream {
+    s_type: StreamType,
```

Comment:
`proto` call it `Kind`. How about aligning them?

Reply:
Aligned :+1: